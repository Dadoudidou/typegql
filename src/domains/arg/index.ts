import { ResolvableType } from 'services/utils';
import { registerArg } from './registry';
export {} from './registry';
export { compileFieldArgs } from './compiler';

// import { getParameterNames } from './services';

interface FieldConfig {
  description?: string;
  type?: ResolvableType;
}

export function Arg(options: FieldConfig = {}): ParameterDecorator {
  return (target: Object, fieldName: string, argIndex: number) => {
    // const allArgNames = getParameterNames(target);
    // const inferedArgName = allArgNames[argIndex];
    registerArg(target.constructor, fieldName, argIndex, { ...options });
  };
}
