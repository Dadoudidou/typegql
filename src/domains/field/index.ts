import { fieldsRegistry, FieldInnerConfig } from './registry';

export { FieldInnerConfig, fieldsRegistry, queryFieldsRegistry } from './registry';
export { compileAllFields, compileFieldConfig } from './compiler';
export { isQueryField } from './services';
export { FieldError } from './error';

export interface FieldOptions {
  description?: string;
  type?: any;
  name?: string;
}

export function Field(options?: FieldOptions): PropertyDecorator {
  return (targetInstance: Object, fieldName: string) => {
    const finalConfig: FieldInnerConfig = {
      property: fieldName,
      name: fieldName,
      ...options,
    };

    fieldsRegistry.set(targetInstance.constructor, fieldName, finalConfig);
  };
}
