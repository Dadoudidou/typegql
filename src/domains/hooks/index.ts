import {
  registerFieldAfterHook,
  registerFieldBeforeHook,
  HookExecutor,
} from './registry';

export {
  fieldAfterHooksRegistry,
  fieldBeforeHooksRegistry,
  HookExecutor,
} from './registry';
export { HookError } from './error';

export function Before(hook: HookExecutor): PropertyDecorator {
  return (targetInstance: Object, fieldName: string) => {
    registerFieldBeforeHook(targetInstance.constructor, fieldName, hook);
  };
}

export function After(hook: HookExecutor): PropertyDecorator {
  return (targetInstance: Object, fieldName: string) => {
    registerFieldAfterHook(targetInstance.constructor, fieldName, hook);
  };
}

export interface GuardOptions {
  msg: string;
}

export function Guard(
  guardFunction: HookExecutor<boolean>,
  options: GuardOptions,
): PropertyDecorator {
  return (targetInstance: Object, fieldName: string) => {
    const hook: HookExecutor = async data => {
      const isAllowed = await guardFunction(data);
      if (isAllowed !== true) {
        throw new Error(options.msg);
      }
    };
    registerFieldBeforeHook(targetInstance.constructor, fieldName, hook);
  };
}

export function createGuard(guardFunction: HookExecutor<boolean>, options: GuardOptions) {
  return (guardOptions: GuardOptions = options): PropertyDecorator => {
    return Guard(guardFunction, guardOptions);
  };
}
