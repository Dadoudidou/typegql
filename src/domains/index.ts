export { Type, compileType } from './type';
export { Field } from './field';
export { Arg } from './arg';
export { Inject, Context, Source, Info } from './inject';
export { registerEnum } from './enum';
