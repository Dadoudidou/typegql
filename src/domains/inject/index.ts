import { registerInjector, InjectorResolver } from './registry';
export {
  getAllInjectors,
  getInjectorResolver,
  hasInjectorRegistered,
  InjectorsIndex,
  InjectorResolver,
} from './registry';

export function Inject(resolver: InjectorResolver): ParameterDecorator {
  return (target: Object, fieldName: string, argIndex: number) => {
    registerInjector(target.constructor, fieldName, argIndex, resolver);
  };
}

export const Context = Inject((source, args, context, info) => {
  return context;
});

export const Info = Inject((source, args, context, info) => {
  return info;
});

export const Source = Inject((source, args, context, info) => {
  return source;
});
