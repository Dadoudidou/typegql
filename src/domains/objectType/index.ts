import { compileObjectTypeWithConfig, compileObjectTypeAsInputWithConfig } from './compiler';
import { objectTypeRegistry, inputTypeRegistry } from './registry';

export { compileObjectType } from './compiler';
export { ObjectTypeError } from './error';
export { objectTypeRegistry, inputTypeRegistry } from './registry';

export interface ObjectTypeOptions {
  name?: string;
  description?: string;
}

export function ObjectType(options?: ObjectTypeOptions): ClassDecorator {
  return (target: Function) => {
    const config = { name: target.name, ...options };
    const outputTypeCompiler = () =>
      compileObjectTypeWithConfig(target, config);
    objectTypeRegistry.set(target, outputTypeCompiler);
    
    const inputTypeCompiler = () =>
    compileObjectTypeAsInputWithConfig(target, config);
    inputTypeRegistry.set(target, inputTypeCompiler);
  };
}
