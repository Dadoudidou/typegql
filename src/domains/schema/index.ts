export { schemaRegistry, mutationFieldsRegistry, queryFieldsRegistry } from './registry';
import { schemaRegistry } from './registry';
import { compileSchema } from './compiler';
export { compileSchema } from './compiler';
export { Query, Mutation } from './rootFields';

export function Schema(): ClassDecorator {
  return target => {
    const compiler = () => compileSchema(target);
    schemaRegistry.set(target, compiler);
  };
}
