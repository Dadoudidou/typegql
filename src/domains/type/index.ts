// import { registerObjectType } from './registry';

export { compileType } from './compiler';
import { typeConfigRegistry } from './registry';
export { typeConfigRegistry } from './registry';
export { ObjectTypeError } from './error';

export const objectTypeSymbol = Symbol('objectType');

declare global {
  interface Function {
    [objectTypeSymbol]: any;
  }
}

export function Type(options?: any): ClassDecorator {
  return (target: Function) => {
    const { name } = target;
    typeConfigRegistry.set(target, { name, ...options });
  };
}
