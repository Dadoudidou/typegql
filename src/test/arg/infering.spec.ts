import { GraphQLString, GraphQLFloat } from 'graphql';
import { Field, Type, compileType } from 'domains';

describe('Arguments', () => {
  it('Infers basic arguments without @Arg decorator', () => {
    @Type()
    class Foo {
      @Field()
      bar(baz: string): string {
        return baz;
      }
    }
    const { bar } = compileType(Foo).getFields();

    expect(bar.args.length).toBeGreaterThan(0);
    const [bazArg] = bar.args;
    expect(bazArg.type).toBe(GraphQLString);
    expect(bazArg.name).toBe('baz');
  });

  it('Throws if is not able to infer arguemtn type without @Arg decorator', () => {
    @Type()
    class Foo {
      @Field()
      bar(baz: any): string {
        return baz;
      }
    }
    expect(() => compileType(Foo).getFields()).toThrowErrorMatchingSnapshot();
  });

  it('Infers multiple basic arguments without @Arg decorator', () => {
    @Type()
    class Foo {
      @Field()
      bar(baz: string, boo: number): string {
        return baz;
      }
    }
    const { bar } = compileType(Foo).getFields();

    expect(bar.args.length).toEqual(2);
    const [bazArg, booArg] = bar.args;
    expect(bazArg.type).toBe(GraphQLString);
    expect(bazArg.name).toBe('baz');
    expect(booArg.name).toBe('boo');
    expect(booArg.type).toBe(GraphQLFloat);
  });
});
