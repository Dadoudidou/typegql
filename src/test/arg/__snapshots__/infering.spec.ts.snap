// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Arguments Throws if is not able to infer arguemtn type without @Arg decorator 1`] = `"@Type Foo.bar(baz <-------): Could not infer type of argument"`;
